src/sync.h

```diff
@@ -400,4 +405,178 @@ class CSemaphoreGrant
     }
 };
 
+/**
+ * A smart container that protects the owned object with a mutex.
+ * Every time the container is dereferenced a mutex is acquired and held locked
+ * for the duration of the method being executed. For example, the following is
+ * safe:
+ * @code{.cpp}
+ * Synced<std::vector<int>> v;
+ *
+ * std::thread t1{[&v]() { v->push_back(5); }};
+ * std::thread t2{[&v]() { v->push_back(6); }};
+ * @endcode
+ *
+ * It is possible so acquire the mutex for longer time than just for a single call:
+ * @code{.cpp}
+ * Synced<std::vector<int>> v{5, 6, 7, 8};
+ *
+ * for (auto& i : v.Lock()) {
+ *     i *= 10;
+ * }
+ *
+ * {
+ *     LOCK(v);
+ *     size_t size = v->size();
+ *     // The container will not change between the two size() calls.
+ *     assert(size == v->size());
+ * }
+ * @endcode
+ */
+template <typename T>
+class Synced
+{
+public:
+    /**
+     * Construct the smart container and its internal object of type T,
+     * passing `args...` to T's constructor.
+     */
+    template <typename... Args>
+    Synced(Args... args) : m_owner{}, m_obj{args...}
+    {
+    }
+
+    /**
+     * An auxiliary class that holds a mutex locked for the duration of its
+     * lifetime and provides access to another object.
+     */
+    class Proxy
+    {
+    public:
+        Proxy(Synced& parent, const char* mutex_name, const char* file_name, int line)
+            // Avoid double-lock if the current thread is already the owner.
+            // This read of parent.m_owner is unprotected but that is ok because
+            // the result of the comparison to the current thread cannot be
+            // changed by other, concurrently executing, threads.
+            : m_lock{parent.m_owner == std::this_thread::get_id() ? nullptr : &parent.m_mutex,
+                     mutex_name,
+                     file_name,
+                     line},
```

Comment:
I agree. Simplified this by removing `m_owner` and using `RecursiveMutex`.

This reduced the `Synced` implementation from 110 to 58 lines (excluding comments).

It is possible to further reduce it by removing `Synced::UniqueLock` but then call sites would have to use `auto lock = foo.Lock();` instead of the sweet `LOCK(foo);`. I think either way is fine.

> Maybe it would be better to disallow double locking and use clang thread annotations to prevent obvious cases of double locking at compile time. But maybe this is not possible.

I will look at this again, but last time I tried I couldn't do that - stumbled on some limitations and bizarre behavior that looked like a bug in the thread safety annotations. Notice that even now some cases are prevented at compile time, like:

```cpp
LOCK(foo);
...
LOCK(foo);
```